### PR TITLE
glib,gtk: correct CLAGS for plain buildtype

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
A little fallout from #58310 but this should correct things to what
I believe they were or should have been.

Noticed this fairly quickly because of https://github.com/NixOS/nixpkgs/issues/65843.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by worldofpeace** [comment]
Guess some discussion might be happening upstream
* https://gitlab.gnome.org/GNOME/glib/issues/1833

**Comment by worldofpeace** [comment]
Looks like we should do `G_DISABLE_CAST_CHECKS` in `gnome-settings-daemon` too
* https://gitlab.gnome.org/GNOME/gnome-settings-daemon/blob/master/meson.build#L74

**Review by 7c6f434c — APPROVED** [review]

File: pkgs/development/libraries/glib/default.nix [hunk 1]
```diff
@@ -98,8 +98,12 @@ stdenv.mkDerivation rec {
...
   LC_ALL = "en_US.UTF-8";
 
-  NIX_CFLAGS_COMPILE = (optional stdenv.isSunOS "-DBSD_COMP")
-    ++ [ "-Wno-error=nonnull" ];
+  NIX_CFLAGS_COMPILE = [
+    "-Wno-error=nonnull"
+    # Default for release buildtype but passed manually because
+    # we're using plain
+    "-DG_DISABLE_CAST_CHECKS"
+  ] ++ optional stdenv.isSunOS "-DBSD_COMP";
```
Comment (jtojnar) [inline pkgs/development/libraries/glib/default.nix]: Any idea what is this for? It originated in https://github.com/NixOS/nixpkgs/commit/8ae2a89cabb83428d5217ac8623b70935317c1b1 but I do not see any mentions in `git log -p -SBSD_COMP` in Glib repository.

**Comment by jtojnar** [comment]
Could you add links to relevant sections of `meson.build` to the other commit messages as well?

Also please describe in the `gtk3` commit message why we need `G_ENABLE_DEBUG`. It will be very helpful in the future, when I will be trying to figure out which flags to prune using `git blame`.

Otherwise, it looks fine.

File: pkgs/development/libraries/glib/default.nix (hunk 1 above)
Comment (worldofpeace) [reply pkgs/development/libraries/glib/default.nix]: Yeah I have no idea either, since we've diverged so much from then I think we can just drop it.

**Comment by worldofpeace** [comment]
> Could you add links to relevant sections of `meson.build` to the other commit messages as well?
> 
> Also please describe in the `gtk3` commit message why we need `G_ENABLE_DEBUG`. It will be very helpful in the future, when I will be trying to figure out which flags to prune using `git blame`.
> 
> Otherwise, it looks fine.

Yep, will do all of that.

**Comment by worldofpeace** [comment]
@jtojnar rewritten/revised

**Review by jtojnar — APPROVED** [review]

**Pull request merged**